### PR TITLE
feat: add hook logic to ManagerWithdrawHook

### PR DESCRIPTION
**Pull request by ramenforbreakfast**
I ultimately decided to remove the `onlyCollateral` modifier to the `hook` here because it is only a view function. We cannot change it to a view function to abide to our interface, but it doesn't matter and the modifier can be added back in if we ever decide to perform write operations.

This also simplifies testing since if `onlyCollateral` was on there, I would need to introduce a wrapper contract for testing purposes. Usually our control modifiers reference something that is only used for checking access, but here the contract itself also uses the `collateral` contract to determine reserve requirements, and you cannot sign a transaction using a contract.

I decided to make `getRequiredReserve` a public function since I think it would be useful to obtain it externally.

### CODE REVIEW

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts [hunk 1]
```diff
@@ -306,4 +306,73 @@ describe('=> ManagerWithdrawHook', () => {
...
+
+  describe('# getRequiredReserve', () => {
+    beforeEach(async () => {
+      await setupManagerWithdrawHook()
+    })
+
+    it('reverts if deposit record not configured', async () => {
+      await managerWithdrawHook.connect(deployer).setDepositRecord(ZERO_ADDRESS)
+
+      await expect(managerWithdrawHook.getRequiredReserve()).reverted
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:318]: nice concise new style - let's use this going forward

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts [hunk 2]
```diff
@@ -306,4 +306,73 @@ describe('=> ManagerWithdrawHook', () => {
...
+    })
+
+    it('returns correct reserve requirement', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:321]:
```suggestion
    it('returns required reserve', async () => {
```

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts [hunk 3]
```diff
@@ -306,4 +306,73 @@ describe('=> ManagerWithdrawHook', () => {
...
+      const requiredReserve = TEST_GLOBAL_DEPOSIT_CAP.mul(TEST_MIN_RESERVE_PERCENTAGE).div(
+        RESERVE_DENOMINATOR
+      )
+      const amountToWithdraw = parseEther('1')
+      collateral.getReserve.returns(requiredReserve.add(amountToWithdraw).add(1))
+      expect(await collateral.getReserve()).to.be.gt(requiredReserve.add(amountToWithdraw))
+
+      await managerWithdrawHook
+        .connect(user)
+        .hook(user.address, amountToWithdraw.add(1), amountToWithdraw)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:375]: i think it's a bit confusing with the 2nd and 3rd arguments - can we make the variable for the ignored argument explicitly be a `IGNORED_ARGUMENT` variable?
same for elsewhere (could also do this in a separate PR)

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts [hunk 4]
```diff
@@ -306,4 +306,73 @@ describe('=> ManagerWithdrawHook', () => {
         .withArgs(TEST_MIN_RESERVE_PERCENTAGE)
     })
   })
+
+  describe('# getRequiredReserve', () => {
+    beforeEach(async () => {
+      await setupManagerWithdrawHook()
+    })
+
+    it('reverts if deposit record not configured', async () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:315]:
```suggestion
    it('reverts if deposit record not set', async () => {
```

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts [hunk 5]
```diff
@@ -306,4 +306,73 @@ describe('=> ManagerWithdrawHook', () => {
...
+      depositRecord.getGlobalNetDepositAmount.returns(TEST_GLOBAL_DEPOSIT_CAP)
+      const requiredReserve = TEST_GLOBAL_DEPOSIT_CAP.mul(TEST_MIN_RESERVE_PERCENTAGE).div(
+        RESERVE_DENOMINATOR
+      )
+      collateral.getReserve.returns(requiredReserve)
+      const amountToWithdraw = 1
+      expect(await collateral.getReserve()).to.be.lt(requiredReserve.add(amountToWithdraw))
+
+      await expect(
+        managerWithdrawHook.connect(user).hook(user.address, amountToWithdraw + 1, amountToWithdraw)
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:346]: shouldnt it just be withdrawing amountToWithdraw? no +1 necessary because already exactly at required reserve beforehand

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts [hunk 6]
```diff
@@ -306,4 +306,73 @@ describe('=> ManagerWithdrawHook', () => {
...
+    })
+
+    it('allows withdrawal down to exactly reserve minimum', async () => {
+      depositRecord.getGlobalNetDepositAmount.returns(TEST_GLOBAL_DEPOSIT_CAP)
+      const requiredReserve = TEST_GLOBAL_DEPOSIT_CAP.mul(TEST_MIN_RESERVE_PERCENTAGE).div(
+        RESERVE_DENOMINATOR
+      )
+      const amountToWithdraw = parseEther('1')
+      collateral.getReserve.returns(requiredReserve.add(amountToWithdraw))
+      expect(await collateral.getReserve()).to.be.gte(requiredReserve.add(amountToWithdraw))
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:357]: confusing setup - can just do the same as above and ensure that reserve returns requiredReserve + 1
then just withdraw 1

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts (hunk 5 above)
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:346]:
```suggestion

      await expect(
        managerWithdrawHook.connect(user).hook(user.address, 1, IGNORED_ARGUMENT)
```

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts (hunk 6 above)
Comment (XavierEkkel) [reply apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:357]: similarly for one below, can just return requiredReserve + 2, then withdraw only 1

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

File: apps/smart-contracts/core/contracts/ManagerWithdrawHook.sol [hunk 7]
```diff
@@ -24,16 +24,16 @@ contract ManagerWithdrawHook is
...
-  }
-
   function hook(
     address _sender,
     uint256 _amountBeforeFee,
     uint256 _amountAfterFee
-  ) external override {}
+  ) external override {
+    require(
+      _collateral.getReserve() >= getRequiredReserve() + _amountAfterFee,
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/ManagerWithdrawHook.sol:33]: nit: clearer way of writing
```suggestion
      _collateral.getReserve() - _amountAfterFee >= getRequiredReserve(),
```

File: apps/smart-contracts/core/contracts/ManagerWithdrawHook.sol [hunk 8]
```diff
@@ -79,4 +79,10 @@ contract ManagerWithdrawHook is
   function getMinReservePercentage() external view override returns (uint256) {
     return _minReservePercentage;
   }
+
+  function getRequiredReserve() public view override returns (uint256) {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/ManagerWithdrawHook.sol:83]:
```suggestion
  function getMinReserve() public view override returns (uint256) {
```
for consistency

File: apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts (hunk 5 above)
Comment (ramenforbreakfast) [reply apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:346]: The + 1 is for the ignored argument, I think this sort of ties in to your previous argument where I should make it more clear using an IGNORED_ARGUMENT variable.

Comment (ramenforbreakfast) [reply apps/smart-contracts/core/test/ManagerWithdrawHook.test.ts:346]: I think you meant to switch the arguments here (we ignore `amountBeforeFee`, not `amountAfterFee`), but understand what you mean.
